Refactor  aspects membership dropdown to use handlebars template

cmrd-senya opened the pull request:
Aspect membership dropdown is used at the contacts, notifications, getting started pages, people hovercards and at a profile page header.

Before there was the [aspect_membership_dropdown](https://github.com/diaspora/diaspora/blob/develop/app/views/aspect_memberships/_aspect_membership_dropdown.haml) template which was used to render the dropdown serverside. Each update to the dropdown required to reload the dropdown using the [controller](https://github.com/diaspora/diaspora/blob/develop/app/controllers/people_controller.rb#L159-L170). Also, there was an AspectMembership backbone view which didn't use templates but rather attached to the DOM generated by serverside rendering.

This PR adds a handlebars template for the aspect_membership_dropdown element and refactors the AspectMembership view for using the template.

I started this job in order to fix #6120. I found out then, that it's not strictly necessary to do all these refactorings to fix #6120, but it is still a nice thing to do.

Currently this PR contains changes for profile, contacts and getting started pages, and notifications page is unchanged. I felt, that finishing this getting complicated, so I felt I need to get some feedback to proceed, to be sure that I'm doing the right thing before diving into more complex and large changes. Here is the list of thing left to do to finish the PR IMO:
- [x] Fix up notifications page to use handlebars template as well
- [x] Fix up peoples' hovercards
- [x] search pages should use it also
- [x] people/:guid/contacts
- [x] Fix up jasmine tests for the AspectMembership view component (there are more async stuff now, so it's not too straightforward)
- [x] Write some tests for the new code
- [x] Wipe everything left from the serverside rendering.

Also, there is a component called [AspectDropdown view](https://github.com/diaspora/diaspora/blob/develop/app/assets/javascripts/app/views/aspects_dropdown_view.js) which previously was a base for the AspectMembership view. The AspectDropdown view is used in the publisher and also renders serverside. This PR makes AspectMembership independent of the AspectDropdown. Maybe it worth making AspectDropdown template based as well and basing AspectMembership on AspectDropdown again since they are definetely related. I can do that as well, but it's some extra work and I'd like to avoid it for a while if it could be separated from this job. However, if you say it's wanted, I could make it.

Commit:
Clear gon after each rspec test

Commit:
Rename app.collections.Aspect{,Selection}s

Commit:
Refactor app.views.AspectMembership

in order to support adding new aspect to a dropdown without full
page reload

diaspora-code-review commented on app/views/aspect_memberships/_aspect_membership_dropdown.haml:
```diff
@@ -9,7 +9,7 @@
     %span.caret
 
-  %ul.dropdown-menu{:class => ["pull-#{hang}", defined?(dropdown_class) && dropdown_class], :unSelectable => 'on', 'data-person_id' => (person.id if defined?(person) && person), 'data-service_uid' => (service_uid if defined?(service_uid)), 'data-person-short-name' => (person.first_name if defined?(person) && person)}
+  %ul.dropdown-menu{:class => ["pull-#{hang}", "aspect_membership"], :unSelectable => 'on', 'data-person_id' => (person.id if defined?(person) && person), 'data-service_uid' => (service_uid if defined?(service_uid)), 'data-person-short-name' => (person.first_name if defined?(person) && person)}
```
Line is too long. [295/120]

diaspora-code-review replied:
Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.

diaspora-code-review commented on app/assets/javascripts/app/collections/aspect_memberships.js:
```diff
@@ -1,6 +1,15 @@
+  model: app.models.AspectMembership,
+
+  findByAspect : function(id) {
+    return this.find(app.collections.AspectMemberships.belongsToAspect, {aspect_id: id});
```
Identifier 'aspect_id' is not in camel case.

diaspora-code-review commented on app/assets/javascripts/app/collections/aspect_memberships.js:
```diff
@@ -1,6 +1,15 @@
+},{
+  belongsToAspect: function(membership) {
+    var aspect = membership.get("aspect");
+    return aspect && aspect.id === this.aspect_id;
```
Identifier 'aspect_id' is not in camel case.

diaspora-code-review commented on app/assets/javascripts/app/collections/aspect_selections.js:
```diff
@@ -0,0 +1,29 @@
+  },
+
+  selectAll: function(){
+    this.map(function(a){ a.set({ 'selected' : true })} );
```
Strings must use doublequote.

diaspora-code-review commented on app/assets/javascripts/app/collections/aspect_selections.js:
```diff
@@ -0,0 +1,29 @@
+  },
+
+  deselectAll: function(){
+    this.map(function(a){ a.set({ 'selected' : false })} );
```
Strings must use doublequote.

diaspora-code-review commented on app/assets/javascripts/app/collections/aspect_selections.js:
```diff
@@ -0,0 +1,29 @@
+  },
+
+  toSentence: function(){
+    var separator = Diaspora.I18n.t("comma") + ' ';
```
Strings must use doublequote.

diaspora-code-review commented on app/assets/javascripts/app/collections/aspect_selections.js:
```diff
@@ -0,0 +1,29 @@
+
+  toSentence: function(){
+    var separator = Diaspora.I18n.t("comma") + ' ';
+    return this.selectedAspects('name').join(separator).replace(/,\s([^,]+)$/, ' ' + Diaspora.I18n.t("and") + ' $1') || Diaspora.I18n.t("my_aspects");
```
Line is too long.

diaspora-code-review replied:
Strings must use doublequote.

diaspora-code-review commented on app/assets/javascripts/app/models/contact.js:
```diff
@@ -7,7 +7,7 @@ app.models.Contact = Backbone.Model.extend({
 
   inAspect : function(id) {
-    return this.aspectMemberships.any(function(membership){ return membership.get("aspect").id === id; });
+    return this.aspectMemberships.any(app.collections.AspectMemberships.belongsToAspect, {aspect_id: id});
```
Identifier 'aspect_id' is not in camel case.

diaspora-code-review commented on app/assets/javascripts/app/models/person.js:
```diff
@@ -8,6 +8,10 @@ app.models.Person = Backbone.Model.extend({
   initialize: function() {
     if( this.get('profile') )
       this.profile = new app.models.Profile(this.get('profile'));
+    if( this.get('contact') ) {
```
Strings must use doublequote.

diaspora-code-review commented on app/assets/javascripts/app/models/person.js:
```diff
@@ -8,6 +8,10 @@ app.models.Person = Backbone.Model.extend({
     if( this.get('profile') )
       this.profile = new app.models.Profile(this.get('profile'));
+    if( this.get('contact') ) {
+      this.contact = new app.models.Contact(this.get('contact'));
```
Strings must use doublequote.

diaspora-code-review commented on app/assets/javascripts/app/models/stream_aspects.js:
```diff
@@ -26,8 +26,8 @@ app.models.StreamAspects = app.models.Stream.extend({
     this.triggerFetchedEvents();
-    if (app.aspects) {
-      app.aspects.trigger('aspectStreamFetched');
+    if (app.aspect_selections) {
```
Identifier 'aspect_selections' is not in camel case.

diaspora-code-review commented on app/assets/javascripts/app/models/stream_aspects.js:
```diff
@@ -26,8 +26,8 @@ app.models.StreamAspects = app.models.Stream.extend({
-    if (app.aspects) {
-      app.aspects.trigger('aspectStreamFetched');
+    if (app.aspect_selections) {
+      app.aspect_selections.trigger('aspectStreamFetched');
```
Identifier 'aspect_selections' is not in camel case.

diaspora-code-review replied:
Strings must use doublequote.

diaspora-code-review commented on app/assets/javascripts/app/pages/getting_started.js:
```diff
@@ -0,0 +1,18 @@
+  },
+
+  aspectMembershipView: function() {
+    return new app.views.AspectMembership({person: this.inviter, dropdown_may_create_new_aspect: true});
```
Identifier 'dropdown_may_create_new_aspect' is not in camel case.

diaspora-code-review commented on app/assets/javascripts/app/router.js:
```diff
@@ -134,14 +143,14 @@ app.Router = Backbone.Router.extend({
   aspects: function() {
-    app.aspects = app.aspects || new app.collections.Aspects(app.currentUser.get("aspects"));
-    this.aspectsList = this.aspectsList || new app.views.AspectsList({ collection: app.aspects });
+    app.aspect_selections = app.aspect_selections || new app.collections.AspectSelections(app.currentUser.get("aspects"));
```
Line is too long.

diaspora-code-review replied:
Identifier 'aspect_selections' is not in camel case.

diaspora-code-review commented on app/assets/javascripts/app/router.js:
```diff
@@ -134,14 +143,14 @@ app.Router = Backbone.Router.extend({
-    app.aspects = app.aspects || new app.collections.Aspects(app.currentUser.get("aspects"));
-    this.aspectsList = this.aspectsList || new app.views.AspectsList({ collection: app.aspects });
+    app.aspect_selections = app.aspect_selections || new app.collections.AspectSelections(app.currentUser.get("aspects"));
+    this.aspectsList = this.aspectsList || new app.views.AspectsList({ collection: app.aspect_selections });
```
Identifier 'aspect_selections' is not in camel case.

diaspora-code-review commented on app/assets/javascripts/app/router.js:
```diff
@@ -134,14 +143,14 @@ app.Router = Backbone.Router.extend({
 
   aspects_stream : function(){
-    var ids = app.aspects.selectedAspects("id");
+    var ids = app.aspect_selections.selectedAspects("id");
```
Identifier 'aspect_selections' is not in camel case.

diaspora-code-review commented on app/assets/javascripts/app/views/aspect_create_view.js:
```diff
@@ -39,27 +37,44 @@ app.views.AspectCreate = app.views.Base.extend({
-      app.events.trigger("aspect:create", aspectId);
+      if(membership) {
+        if(!this.person.contact)
+          this.person.contact = new app.models.Contact();
```
Expected '{' and instead saw 'this'.

diaspora-code-review commented on app/assets/javascripts/app/views/aspect_create_view.js:
```diff
@@ -39,27 +37,44 @@ app.views.AspectCreate = app.views.Base.extend({
+  ensureEventsOrder: function() {
+    this._eventsCounter++;
+    if(this._eventsCounter > 1)
+      app.events.trigger("aspect:create", this._newAspectId);
```
Expected '{' and instead saw 'app'.

diaspora-code-review commented on app/assets/javascripts/app/views/aspect_membership_view.js:
```diff
@@ -18,70 +24,87 @@ app.views.AspectMembership = app.views.AspectsDropdown.extend({
+  },
+
+  presenter: function() {
+    aspect_memberships_length = this.person.contact ? this.person.contact.aspectMemberships.length : 0;
```
Identifier 'aspect_memberships_length' is not in camel case.

diaspora-code-review commented on app/assets/javascripts/app/views/aspect_membership_view.js:
```diff
@@ -18,70 +24,87 @@ app.views.AspectMembership = app.views.AspectsDropdown.extend({
+    aspect_memberships_length = this.person.contact ? this.person.contact.aspectMemberships.length : 0;
+
+    return _.extend(this.defaultPresenter(), {
+      aspect_memberships_length: aspect_memberships_length,
```
Identifier 'aspect_memberships_length' is not in camel case.

diaspora-code-review commented on app/assets/javascripts/app/views/aspect_membership_view.js:
```diff
@@ -18,70 +24,87 @@ app.views.AspectMembership = app.views.AspectsDropdown.extend({
+    return _.extend(this.defaultPresenter(), {
+      aspect_memberships_length: aspect_memberships_length,
+      aspects: this.aspectsPresenter(),
+      dropdown_may_create_new_aspect: this.dropdown_may_create_new_aspect,
```
Identifier 'dropdown_may_create_new_aspect' is not in camel case.

diaspora-code-review commented on app/assets/javascripts/app/views/aspect_membership_view.js:
```diff
@@ -18,70 +24,87 @@ app.views.AspectMembership = app.views.AspectsDropdown.extend({
+      aspect_memberships_length: aspect_memberships_length,
+      aspects: this.aspectsPresenter(),
+      dropdown_may_create_new_aspect: this.dropdown_may_create_new_aspect,
+    }, aspect_memberships_length == 0 ? {
```
Identifier 'aspect_memberships_length' is not in camel case.

diaspora-code-review replied:
Expected '===' and instead saw '=='.

diaspora-code-review commented on app/assets/javascripts/app/views/aspect_membership_view.js:
```diff
@@ -18,70 +24,87 @@ app.views.AspectMembership = app.views.AspectsDropdown.extend({
+      aspects: this.aspectsPresenter(),
+      dropdown_may_create_new_aspect: this.dropdown_may_create_new_aspect,
+    }, aspect_memberships_length == 0 ? {
+      extra_button_class: ""
```
Identifier 'extra_button_class' is not in camel case.

diaspora-code-review commented on app/assets/javascripts/app/views/aspect_membership_view.js:
```diff
@@ -18,70 +24,87 @@ app.views.AspectMembership = app.views.AspectsDropdown.extend({
+    }, aspect_memberships_length == 0 ? {
+      extra_button_class: ""
+    } : { // this.contact.aspectMemberships.length > 0
+      all_aspects_are_selected: aspect_memberships_length == app.aspects.length,
```
Identifier 'all_aspects_are_selected' is not in camel case.

diaspora-code-review replied:
Identifier 'aspect_memberships_length' is not in camel case.

diaspora-code-review replied:
Expected '===' and instead saw '=='.

diaspora-code-review commented on app/assets/javascripts/app/views/aspect_membership_view.js:
```diff
@@ -18,70 +24,87 @@ app.views.AspectMembership = app.views.AspectsDropdown.extend({
+      extra_button_class: ""
+    } : { // this.contact.aspectMemberships.length > 0
+      all_aspects_are_selected: aspect_memberships_length == app.aspects.length,
+      only_one_aspect_is_selected: aspect_memberships_length == 1,
```
Identifier 'only_one_aspect_is_selected' is not in camel case.

diaspora-code-review replied:
Identifier 'aspect_memberships_length' is not in camel case.

diaspora-code-review replied:
Expected '===' and instead saw '=='.

diaspora-code-review commented on app/assets/javascripts/app/views/aspect_membership_view.js:
```diff
@@ -18,70 +24,87 @@ app.views.AspectMembership = app.views.AspectsDropdown.extend({
+    } : { // this.contact.aspectMemberships.length > 0
+      all_aspects_are_selected: aspect_memberships_length == app.aspects.length,
+      only_one_aspect_is_selected: aspect_memberships_length == 1,
+      first_membership_name: this.person.contact.aspectMemberships.at(0).get("aspect").name,
```
Identifier 'first_membership_name' is not in camel case.

diaspora-code-review commented on app/assets/javascripts/app/views/aspect_membership_view.js:
```diff
@@ -18,70 +24,87 @@ app.views.AspectMembership = app.views.AspectsDropdown.extend({
+      all_aspects_are_selected: aspect_memberships_length == app.aspects.length,
+      only_one_aspect_is_selected: aspect_memberships_length == 1,
+      first_membership_name: this.person.contact.aspectMemberships.at(0).get("aspect").name,
+      extra_button_class: "btn-success"
```
Identifier 'extra_button_class' is not in camel case.

denschub commented:
> POST https://api.github.com/repos/diaspora/diaspora/pulls/6864/comments: 403 - You have triggered an abuse detection mechanism and have been temporarily blocked from content creation. Please retry your request again later

Good work getting our stylecheck robot spamblocked by Github! 👍

cmrd-senya commented:
My apologize :(

diaspora-code-review commented on app/assets/javascripts/app/views/aspect_membership_view.js:
```diff
@@ -18,70 +24,89 @@ app.views.AspectMembership = app.views.AspectsDropdown.extend({
+    });
 
-    return aspect_membership.save();
+    return this.person.contact.aspectMemberships.create({aspect_id: aspectId, person_id: this._personId()});
```
Identifier 'aspect_id' is not in camel case.

diaspora-code-review replied:
Identifier 'person_id' is not in camel case.

diaspora-code-review commented on spec/javascripts/app/collections/aspect_selections_spec.js:
```diff
@@ -0,0 +1,94 @@
+    var locale = {
+      and: "and",
+      comma: ",",
+      my_aspects: "My Aspects"
```
Identifier 'my_aspects' is not in camel case.

diaspora-code-review commented on app/assets/stylesheets/aspects.scss:
```diff
@@ -29,4 +22,13 @@
+  li.selected {
+    .glyphicon-ok { display: inline-block;}
+    .glyphicon-refresh { display: none;}
+  }
```
Rule declaration should be followed by an empty line

diaspora-code-review commented on app/assets/stylesheets/aspects.scss:
```diff
@@ -29,4 +22,13 @@
     }
   }
+
+  li.selected {
```
Merge rule `li.selected` with rule on line 3

diaspora-code-review replied:
Avoid qualifying class selectors with an element.

diaspora-code-review commented on app/assets/stylesheets/aspects.scss:
```diff
@@ -29,4 +22,13 @@
+    .glyphicon-ok { display: inline-block;}
+    .glyphicon-refresh { display: none;}
+  }
+  li.loading {
```
Merge rule `li.loading` with rule on line 3

diaspora-code-review replied:
Avoid qualifying class selectors with an element.

svbergerem commented on app/assets/javascripts/app/collections/aspect_memberships.js:
```diff
@@ -1,6 +1,15 @@
-  model: app.models.AspectMembership
+  model: app.models.AspectMembership,
+
+  findByAspect: function(id) {
```
`findByAspectId`?

svbergerem commented on app/assets/javascripts/app/collections/aspect_selections.js:
```diff
@@ -0,0 +1,30 @@
+app.collections.AspectSelections = Backbone.Collection.extend({
+  model: app.models.AspectSelection,
+
+  selectedAspects: function(attribute) {
```
I'd like to see this method renamed. Maybe `selectedAspectValues`?

cmrd-senya replied:
I don't really feel we need the "Aspects" word in that name, because it is in the `AspectSelections` class. `selected` here implicitly means that it is about apsects. And since the method gets attribute from selected aspects, I would call it `getSelectedAttribute` or `selectedGetAttribute` or just `selectedAttribute`.

svbergerem commented on app/assets/javascripts/app/models/person.js:
```diff
@@ -8,6 +8,10 @@ app.models.Person = Backbone.Model.extend({
   initialize: function() {
     if( this.get('profile') )
       this.profile = new app.models.Profile(this.get('profile'));
```
Please also wrap that block in curly braces. I know that you didn't touch this part of the code but you added the new if block below and now it became even more confusing.

svbergerem commented on app/assets/javascripts/app/views.js:
```diff
@@ -65,7 +65,8 @@ app.views.Base = Backbone.View.extend({
       var view = _.isFunction(self[property]) ? self[property]() : self[property];
-      if(view) {
+      if (view && self.$(selector).length > 0) {
+        self.$(selector).first().empty();
```
Shouldn't `.html(...)` in the next line replace the old contents with new ones? Why do we have to empty it first?

cmrd-senya replied:
Unfortunately I don't remember the exact reason I added this line. However, I checked it right now and I see it changes the behavior on the getting started page. I'll study what's going on here and report.

cmrd-senya replied:
Ok, so when we run render for a second time, in case of `app.views.AspectMembership` `selector` matches to `.aspect_membership_dropdown` twice: to the origianal placeholder and to the `div` element generated by a previous render call (which is inside the first placeholder `div`). Since `html(...)` applies to every matched element, it applies to both elements. DOM looks ok after that, but for some reason further `render` calls don't update it properly anymore. Maybe `html(...)` call to nested elements produces some kind of undefined behaviour.

![2016-07-14 15 59 57](https://cloud.githubusercontent.com/assets/10187586/16840243/43673620-49dc-11e6-8819-cdbe5336692f.png)


So while I can fix it by more precise specification for the selector, it worth leaving explicit erase for the case when selector is specified so resulting in nested elements matches. Having explicit erase would make the difference so the next `.html(...)` call would apply only to the top level matched elements (all nested are erased and therefore don't match anymore). I guess this way it is safer.

cmrd-senya replied:
@svbergerem, I managed to write a test that shows the issue that is solved by this line and that fails without it. I shaped it as a separate commit within the PR: https://github.com/diaspora/diaspora/pull/6864/commits/b532754c87581001827161366acdeb99c2bcdaf9

svbergerem commented on spec/javascripts/app/collections/aspect_selections_spec.js:
```diff
@@ -0,0 +1,94 @@
+      and: "and",
+      comma: ",",
+      my_aspects: "My Aspects"
+    };
```
You shouldn't need this after rebasing this PR. We now always load all locales.

svbergerem commented on spec/javascripts/app/collections/aspect_selections_spec.js:
```diff
@@ -0,0 +1,94 @@
+      {name: "Acquaintances", selected: false}
+    ];
+
+    Diaspora.I18n.load(locale);
```
See above.

Commit:
AspectSelections style fixes

Flaburgan commented:
@cmrd-senya this could maybe solve #3771 in the same time?

Commit:
Fix views.js corner case

cmrd-senya commented:
Hey, guys, isn't there aspects dropdown expected in the notifications dropdown? The DOM for it is in place, but there is no visible element. Was it broken unintentionally or  hidden for a reason?

![2016-07-14 19 19 18](https://cloud.githubusercontent.com/assets/10187586/16847165/4ede2c86-49f8-11e6-9896-85bee5c82484.png)

Commit:
Use the factory in the publisher_view_spec.js

Commit:
Client-side rendering of aspect dropdown for hovercards

svbergerem commented:
@cmrd-senya 
> isn't there aspects dropdown expected in the notifications dropdown

AFAIR we didn't have the dropdown there before and since #5707 we use the same partial for notifications in the notifications view and in the notification dropdown. Since we didn't have an aspect membership dropdown there before we just don't display it. (https://github.com/diaspora/diaspora/pull/5707/files#diff-1d3b7aaf9a68499c7fa63be5194906caR179)

I guess the reason for this is the limited width of the notifications dropdown.

cmrd-senya commented:
@svbergerem, I see, so I guess I don't need to support the aspects dropdown in the notifications dropdown.

svbergerem commented:
@cmrd-senya Sounds good to me.

SuperTux88 commented on app/assets/javascripts/app/views/hovercard_view.js:
```diff
@@ -94,14 +95,20 @@ app.views.Hovercard = app.views.Base.extend({
   _populateHovercard: function() {
     var href = this.href();
-    href += "/hovercard.json";
+    href += ".json";
```
Are you sure you want to use the big person-json for the hovercards? Because this also contains the `bio` (and some other unused fields) and can be many times bigger than the hovercard.json.

cmrd-senya replied:
That's true, I wanted opinions on this topic. The main point of this change was to remove HovercardPresenter. I can introduce hovercard presenting method for PersonPresenter that strips bio and other fields and bring the `/hovercard.json` route back.

SuperTux88 replied:
Why do you want to remove the HovercardPresenter?

cmrd-senya replied:
Because `HovercardPresenter` is a presenter for the person model. I feel it's more presice to make a hovercard method for the `PersonPresenter` rather than having a whole separate presenter. Hovercard data is just another presentation for person data, and not some separate instance.

Also the hovercard presenter is not a child of BasePresenter. So even if leave it, it should be refactored heavily. That what I did in the first place, but then I changed my mind.

denschub replied:
I agree with @cmrd-senya, having a `Hovercard` method inside the `PersonPresenter` feels nicer for me.

Commit:
Notifications and search page backend updates

Updates introduce support for preloading contacts to Gon in order
to support client-side rendering of aspect membership dropdown box.

Commit:
Introduce NotificationSerializer

And remove the note_html property from the model.

Commit:
Notifications and search page frontend updates

Updates introduce usage of client-side aspect dropdown rendering

Commit:
Fix mobile getting_started use wrong person object

Commit:
Client-side aspect dropdown for the people/:guid/contacts page

cmrd-senya commented:
Ok, this is for the review.

jhass commented on app/controllers/aspects_controller.rb:
```diff
@@ -14,10 +14,15 @@ def create
       if aspecting_person_id.present?
         connect_person_to_aspect(aspecting_person_id)
+        if @contact
+          result[:aspect_membership] = AspectMembershipPresenter.new(@contact.aspect_memberships.last).base_hash
```
`.last`/`first` without an explicit `order` is not deterministic iirc. Is this supposed to return the one created in line 19? Perhaps that should return it?

jhass commented on app/helpers/contacts_helper.rb:
```diff
@@ -9,4 +9,11 @@ def start_a_conversation_link(aspect, contacts_size)
+
+  def gon_load_contact(contact)
+    Gon.preloads[:contacts_collection] ||= []
+    if contact.id.nil? || !Gon.preloads[:contacts_collection].map {|contact| contact[:id] }.include?(contact.id)
```
What's `contact.id.nil?` supposed to check for, in what situation does that happen?

For the second part, `Gon.preloads[:contacts_collection].none? {|stored_contact| stored_contact[:id] == contact.id }`

I'm also not too found of things like `collection`, `map`, `array`, `hash` in the name of things, can we just call it `contacts` or `user_contacts`?

cmrd-senya replied:
> What's contact.id.nil? supposed to check for, in what situation does that happen?

In [this place](https://github.com/diaspora/diaspora/pull/6864/files#diff-37afe88261f69735dbcf6835afc2718aR4) in case when contact is `new` it presents a hash with `"id":"null"` value. In this case UI is informed that there is no contact for the person and we still have `"person"` field in the preloaded presented hash which is required for UI components to render.

jhass replied:
So you want to check for a `Contact.new`, that's what [`persisted?`](http://api.rubyonrails.org/classes/ActiveRecord/Persistence.html#method-i-persisted-3F) is for

cmrd-senya replied:
> I'm also not too found of things like collection, map, array, hash in the name of things, can we just call it contacts or user_contacts?

Then I'll have to rename [this preload](https://github.com/diaspora/diaspora/blob/develop/app/controllers/people_controller.rb#L81-L83) to, perhaps, `contact_count`. Are you ok with that?

jhass replied:
Defining that one as `contacts_count` seems fair, yes.

jhass commented on app/presenters/contact_presenter.rb:
```diff
@@ -12,6 +12,12 @@ def full_hash
+  private
+
+  def person_without_contact
+    PersonPresenter.new(person, current_user).full_hash.tap {|hash| hash.delete(:contact) }
```
ActiveSupport to the resuce, `.except(:contact)`.

jhass commented on app/presenters/person_presenter.rb:
```diff
@@ -22,6 +22,10 @@ def as_json(_options={})
   end
 
+  def hovercard
+    full_hash.merge(profile: ProfilePresenter.new(profile).public_hash)
```
The old `HovercardPresenter` returned a lot less attributes, no?

cmrd-senya replied:
Well, yes, there are some unnecessary here. I can clean it up a little.

jhass commented on app/views/notifications/_notification.haml:
```diff
@@ -1,9 +1,12 @@
+  - if note.type == "Notifications::StartedSharing" && (!defined?(no_aspect_dropdown) || !no_aspect_dropdown)
+    - contact = current_user.contact_for(note.target)
+    - if contact
+      - gon_load_contact(contact)
```
This feels like too much logic in the view still, perhaps `gon_load_contact` should handle being passed a `Person`?

cmrd-senya replied:
`Notification` has `recipient` attriubute. Perhaps then implement `contact` method for `Notification::StartedSharing`:
```ruby
def contact
  recipient.contact_for(target)
end
```

?

jhass commented on app/views/people/_relationship_action.haml:
```diff
@@ -1,7 +1,8 @@
   - contact ||= Contact.new(:person => person)
-  = aspect_membership_dropdown(contact, person, 'right')
+  - gon_load_contact(contact)
+  .aspect_membership_dropdown.placeholder{data: {person_id: contact.person_id}}
```
Is this partial used too diversely for that logic to move to a controller? Refactoring it into a view helper would be great anyhow I guess.

cmrd-senya replied:
Looks like we have the same [method](https://github.com/diaspora/diaspora/blob/develop/app/controllers/people_controller.rb#L193) in controller that is called in order to generate arguments for the partial. The logic can be moved there.

Commit:
Rename contacts[:count] to contacts_count in gon

cmrd-senya commented:
I fixed all the points by @jhass so far. Anything else?

jhass commented on app/helpers/contacts_helper.rb:
```diff
@@ -9,4 +9,11 @@ def start_a_conversation_link(aspect, contacts_size)
+
+  def gon_load_contact(contact)
+    Gon.preloads[:contacts] ||= []
+    if Gon.preloads[:contacts].none? {|stored_contact| stored_contact[:person][:id] == contact.person.id }
```
`contact.person_id`, unlikely not already loaded here, but just in general, don't load an association just to get at its id.

jhass commented:
Minor note, Ruby side looks good to me. I didn't review any of the JS parts.

svbergerem commented on app/assets/javascripts/app/views/aspect_membership_view.js:
```diff
@@ -18,70 +22,89 @@ app.views.AspectMembership = app.views.AspectsDropdown.extend({
+      aspects: this.aspectsPresenter(),
+      dropdownMayCreateNewAspect: this.dropdownMayCreateNewAspect
+    }, aspectMembershipsLength === 0 ? {
+      extraButtonClass: "",
```
This should be `btn-default`.

svbergerem commented:
Avatars are missing on the contacts page:

![contacts page missing avatars](https://cloud.githubusercontent.com/assets/3798614/17465709/a2d1f34e-5cff-11e6-959a-e498942fd4b7.png)

svbergerem commented:
@cmrd-senya I opened #6945 while reviewing this PR. IIRC there were three lines that would have to be changed. So this is just a reminder to be careful in case of rebases.

svbergerem commented on app/assets/templates/aspect_membership_dropdown_tpl.jst.hbs:
```diff
@@ -0,0 +1,53 @@
+      class="aspect_selector"
+    {{/if}}
+
+    tabindex="0"
```
I guess you only want to add this to the button. (which you already did above)

svbergerem commented on app/controllers/people_controller.rb:
```diff
@@ -78,9 +80,7 @@ def show
         gon.preloads[:photos] = {
           count: Photo.visible(current_user, @person).count(:all)
         }
```
For the sake of unification: could you also change this like you did that for the `contacts_count`? This would also include a change in the profile header view.

svbergerem commented on app/controllers/people_controller.rb:
```diff
@@ -140,9 +140,7 @@ def contacts
           gon.preloads[:photos] = {
             count: Photo.visible(current_user, @person).count(:all)
           }
```
And here.

svbergerem commented on app/controllers/photos_controller.rb:
```diff
@@ -30,9 +30,7 @@ def index
           gon.preloads[:photos] = {
             count: Photo.visible(current_user, @person).count(:all)
           }
```
And here.

cmrd-senya replied:
I would make a helper method for this then. How about create `photos_helper.rb` and put the method in it?

svbergerem commented on spec/javascripts/app/collections/aspect_selections_spec.js:
```diff
@@ -0,0 +1,87 @@
+  });
+
+  describe("#toSentence", function() {
+    describe("without aspects", function() {
```
You also have an aspect here but it's not selected. Maybe merge this with `describe("with one aspect", ...` below and just select the aspect in the second test.

svbergerem commented on spec/javascripts/app/collections/aspect_selections_spec.js:
```diff
@@ -0,0 +1,87 @@
+        this.aspects = new app.collections.AspectSelections([{name: "Work", selected: false}]);
+      });
+
+      it("returns the name of the aspect", function() {
```
`"returns 'My aspects'"`

svbergerem commented:
Apart from my comments the frontend/JavaScript part looks good to me.

Commit:
AspectSelections spec corrections

Commit:
Rename photos[:count] to photos_count in gon

cmrd-senya commented:
Squashed, rebased, green 😎

svbergerem commented:
Thank you! :tada:

cmrd-senya commented:
Thanks everyone for participation, reviews :)
